core/l1_gas.go

```diff
@@ -0,0 +1,106 @@
+package core
+
+import (
+	gomath "math"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+var big10 = big.NewInt(10)
+
+var (
+	// TODO: make configurable
+	L1BaseFeeSlot = common.BigToHash(big.NewInt(2))
+	OverheadSlot  = common.BigToHash(big.NewInt(3))
+	ScalarSlot    = common.BigToHash(big.NewInt(4))
+	DecimalsSlot  = common.BigToHash(big.NewInt(5))
+)
+
+// calculateL1GasUsed returns the gas used to include the transaction data in
+// the calldata on L1
+func calculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	var zeroes uint64
+	var ones uint64
+	for _, byt := range data {
+		if byt == 0 {
+			zeroes++
+		} else {
+			ones++
+		}
+	}
+
+	zeroesGas := zeroes * params.TxDataZeroGas
+	onesGas := (ones + 68) * params.TxDataNonZeroGasEIP2028
+	l1Gas := new(big.Int).SetUint64(zeroesGas + onesGas)
+	return new(big.Int).Add(l1Gas, overhead)
```

Comment:
I think geth has a function somewhere to compute the intrinsic cost of data, which does the exact same thing? Let's use that instead of creating a new function for it.

Reply:
`IntrinsicGas` isn't quite right. It also includes access lists + a fixed overhead (depends if contract creation or not). It also only looks at `tx.Data()`. We look at `tx.Data()`, but then add the overhead for the rest of the transaction's bytes.